### ISSUE TITLE
`deprecated` decorator with multi line reason string doesn't work on implementation when function has overloads

### ISSUE THREAD

**Issue by DetachHead** [microsoft/pyright#5966]
related: #5964

```py
from typing import overload

from typing_extensions import deprecated


@overload
def foo(value: int): ...


@overload
def foo(value: str): ...


@deprecated(
    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    " aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
)
def foo(value: object): ...


foo(1)  # no deprecation warning


@deprecated(
    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    " aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
)
def bar(value: object): ...


bar(1)  # deprecation warning


@overload
def baz(value: int): ...


@overload
def baz(value: str): ...


@deprecated("aa")
def baz(value: object): ...


baz(1)  # deprecation warning
```

pylance 2023.9.12

**Comment by erictraut** [comment]
Yes, you must use a string literal as the argument to `deprecated`, not a dynamic expression such as chained string literals, string literals that are added with a `+` operator, f-strings, etc. If multiple strings are needed, I recommend using triple quotes.

**Comment by DetachHead** [comment]
@erictraut in that case why does it work on functions without overloads?

![image](https://github.com/microsoft/pyright/assets/57028336/b382ea4f-d3bf-4692-9374-34e54ce09c3f)


i don't see why it should be treated any differently when multi-line concatenated strings are still `Constant`s in the ast and are represented in the exact same way:

![image](https://github.com/microsoft/pyright/assets/57028336/0e216b8e-9e1b-46f9-be4f-d091e5404b00)

**Comment by erictraut** [comment]
If you use the `deprecated` feature correctly, the feature will work. If you use it incorrectly, it won't work.

**Comment by KotlinIsland** [comment]
> If you use the deprecated feature correctly, the feature will work. If you use it incorrectly, it won't work.

Responding with pure snark is extremely off-putting and against any notion of collaboration or open-source spirit.

> dynamic expression such as chained string literals

### [String literal concatenation](https://docs.python.org/3/reference/lexical_analysis.html#string-literal-concatenation)

> Multiple adjacent string or bytes literals (delimited by whitespace), possibly using different quoting conventions, are allowed, and their meaning is the same as their concatenation. Thus, "hello" 'world' is equivalent to "helloworld".

### [PEP 702 ](https://peps.python.org/pep-0702/#specification)
> The message must be a string literal.

```py
import ast

ast.parse('"a" "b"')  # Constant("ab")
ast.parse('"a" + "b"')  # BinOp(Constant("a"), Add(), Constant("b"))
```

```py
message = "ab"
@deprecated(message)
def f0(): ...

@deprecated("a" "b")
def f1(): ...

@overload
def f2(): ...
@overload
def f2(x: int): ...
@deprecated("ab")
def f2(x: ): ...

@overload
def f3(value: int): ...
@overload
def f3(value: str): ...
@deprecated("a" "b")
def f3(value: object): ...

f0()  # reported deprecated without a message
f1()  # reported deprecated without a message
f2()  # reported deprecated "ab"
f3()  # no deprecation warning
```

> you must use a string literal as the argument to deprecated

The deprecation annotation for `f0` is being passed a variable, and pyright is not showing any error that this is an invalid usage.

It seems like all the features in pyright are intentionally 'designed' to not be usable correctly. If you have designed this feature to be inconsistent and unintuitive, then it should be documented as such.

Even if you were right about string literal concatenation being dynamic expressions, this feature, as it it is currently implemented(designed?), is inconsistent.

Please update pyright to consistently handle this scenario:
- Show an error when the value passed to `deprecated` is not a string literal
OR
- Show a deprecation warning when an overloaded function is deprecated without a string literal.

And additionally:
- Update pyright to correctly understand that string literal concatenation is equivalent to a string literal.

**Comment by erictraut** [comment]
I agree that the inconsistency between overloaded and non-overloaded functions is a bug. A fix for this will be included in the next release.

I don't plan to add support for concatenated strings or other dynamic string expressions passed to `@deprecated`, at least not at this time. If you want your string to appear in the diagnostic output, you should use a simple string literal.

I also don't plan to add additional errors. Adding errors in this case would be inconsistent with the way pyright (and in most cases mypy) handles dynamic value expressions in similar locations. For example, within `NamedTuple` definitions, `dataclass` decorator calls, base class lists within class definitions, arguments to `isinstance`. Pyright allows dynamic expressions in these locations and degrades gracefully. There are legitimate reasons why someone might want to use a dynamic expressions in value expressions. If a type checker emits errors in these cases, it would be a false positive.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.328, which I just published. It will also be included in a future release of pylance.
